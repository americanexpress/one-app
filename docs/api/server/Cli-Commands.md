[üëà Return to Overview](../README.md)

# CLI Commands

`one-app` Server consists of [NPM Script Commands](https://docs.npmjs.com/misc/scripts) that determine its runtime configuration and behaviors. 

```sh
npm run <one-app-npm-command> [--] [--some-flag]
# e.g. npm run test:unit -- --watch
```

We describe each command and a description of its usage below.

**Contents**
* Install Commands
  * [`preinstall`](#install-commands)
  * [`postinstall`](#install-commands)
* Development Commands
  * [`serve-module`](#test-commands)
  * [`drop-module`](#test-commands)
  * [`set-middleware`](#test-commands)
  * [`set-dev-endpoints`](#test-commands)
* Start Commands
  * [`start`](#start-commands)
  * [`start:watch`](#start-commands)
  * [`prestart:prod-sample`](#start-commands)
  * [`start:prod-sample`](#start-commands)
* Build Commands
  * [`build`](#build-commands)
  * [`build:server`](#build-commands)
  * [`build:bundle`](#build-commands)
  * [`build:artifacts:statics`](#build-commands)
  * [`build:prod-sample`](#build-commands)
  * [`build:sample-modules`](#build-commands)
* Clean Commands
  * [`clean`](#clean-commands)
  * [`clean:build`](#clean-commands)
  * [`clean:test`](#clean-commands)
* Test Commands
  * [`test`](#test-commands)
  * [`test:unit`](#test-commands)
  * [`test:lint`](#test-commands)
  * [`pretest:integration`](#test-commands)
  * [`test:integration`](#test-commands)
  * [`test:danger`](#test-commands)
  * [`test:git-history`](#test-commands)
  * [`test:lockfile`](#test-commands)

## Install Commands

| Command       | Description                                                                                    |
|---------------|------------------------------------------------------------------------------------------------|
| `preinstall`  | Runs `check-engines` package to ensure a compatible Node version is used before `npm install`. |
| `postinstall` | Runs a [`build`](#build-commands) after `npm install`.                                         |

## Development Commands

> üí¨ These commands are for Development only

| Command | Description |
|------|-------------|
| `serve-module` | Accepts a path to a [Holocron Module] and adds it to the local [development CDN]. *See [`serve-module` Arguments](#serve-module-arguments).*  |
| `drop-module` | Accepts a [Holocron Module] name and removes the Module from the local [development CDN]. *See [`drop-module` Arguments](#drop-module-arguments).* |
| `set-middleware` | Accepts a file path to JS file that defines custom [Express Middleware](http://expressjs.com/en/guide/writing-middleware.html#writing-middleware-for-use-in-express-apps) for the [development proxy]. *See [`set-middleware` Arguments](#set-middleware-arguments).* |
| `set-dev-endpoints`| Accepts a Node file containing development endpoints for the [development proxy]. *See [`set-dev-endpoints` Arguments](#set-dev-endpoints-arguments).* |

### `serve-module` Arguments

```sh
npm run serve-module <path-to-module-folder>...
# e.g. npm run serve-module ../../my-module
```

### `drop-module` Arguments

```sh
npm run drop-module <module-name>...
# e.g. npm run drop-module my-module
```

### `set-middleware` Arguments

```sh
npm run set-middleware <path-to-middleware-js>
# e.g. npm run set-middleware ../dev.middleware.js
```

The contents of the [Express Middleware] JS file should match the following shape:

**Shape**

```js
// (app: ExpressApp) => (req: ExpressRequest, res: ExpressResponse, next: ExpressNext) => undefined
module.exports = (app) => (req, res, next) => {
  next();
};
```

### `set-dev-endpoints` Arguments

```sh
npm run set-dev-endpoints <path-to-endpoints-js>...
# e.g. npm run set-dev-endpoints ../dev.endpoints.js
```

The contents of the endpoints JS file should match the following shape:

**Shape**

```js
module.exports = {
  someIdReference: {
    devProxyPath: String, // String local URL path
    destination: String, // String of Destination Hostname
  },
  // ... More entries of the same shape
};
```

## Start Commands

| Command       | Description                                                                                    |
|---------------|------------------------------------------------------------------------------------------------|
| `start`  | Runs `one-app` Server and accepts the [flags listed below](#start-usage). |
| `start:watch` | Executes `one-app` Server and watches Server source files with [`nodemon`](https://nodemon.io/). When source changes, `one-app` reloads with the `npm start`. This command [accepts all the flags as `start`](#start-usage). |
| `prestart:prod-sample` | Runs [`build:sample-modules`](#build-commands) before `start:prod-sample`. |
| `start:prod-sample` | Runs a [Docker Compose](https://docs.docker.com/compose/compose-file/) file to start up the [Integration Test Suite] containers. |


## Build Commands

| Command       | Description                                                                                    |
|---------------|------------------------------------------------------------------------------------------------|
| `build` | Using the [`concurrently`] package, runs `build:server` and `build:bundle`. |
| `build:server` | Executes [Babel] to transpile ES6 Server source files |
| `build:bundle` | Executes [One App Bundler] to build Browser assets |
| `build:artifacts:statics` | Runs [Static Assets Script](../../../scripts/build-static-assets-artifact.js) for [Integration Tests] |
| `build:prod-sample` | Runs `build:sample-modules` and runs the [Build One App Docker Setup Script](../../../scripts/build-one-app-docker-setup.js) for [Integration Tests] |
| `build:sample-modules` | Builds the [Holocron Modules] in [`prod-sample/sample-modules`] folder for [Integration Tests]. *See [`build:sample-modules` Usage](#buildsample-modules-usage).* |

### `build:sample-modules` Usage

```sh
npm run build:sample-modules [--] [--archive-built-artifacts]
# e.g. npm run build:sample-modules -- --archive-built-artifacts
```

#### Flags

| Flag       | Description                                                                                    |
|---------------|------------------------------------------------------------------------------------------------|
| `--archive-built-artifacts`  | Runs `tar` and `gzip` on the the built sample [Holocron Modules] bundles. *Used in [Integration Test Suite]* |


### Start Usage

```sh
[ONE_CONFIG_ENV=<runtime-level>] npm run start (--) (--root-module-name=<root-module-name>) [--module-map-url=<module-map-url>] [--use-middleware]
# e.g. ONE_CONFIG_ENV=development npm run start -- --root-module-name=my-root-module
```

#### Flags

| Flag       | Description                                                                                    |
|---------------|------------------------------------------------------------------------------------------------|
| `--root-module-name`  | Name of the [Holocron Module] that serves as the entry point to your application. |
| `--module-map-url`  | Remote Module Map URL for [`one-app-dev-cdn`] to proxy. *For Development.* |
| `--use-middleware` | Apply a custom middleware configuration for [`one-app-dev-proxy`]. *For Development.* |

#### Environment Variables

Please see the [Environment Variables](./Environment-Variables.md) API docs.

## Clean Commands

| Command       | Description                                                                                    |
|---------------|------------------------------------------------------------------------------------------------|
| `clean`  | Executes `clean:build` and `clean:test`. |
| `clean:build` | Removes folders and files related to [Babel] and [Webpack] builds. |
| `clean:test` | Removes [Jest] Integration Test Report folder. |

## Test Commands

| Command       | Description                                                                                    |
|---------------|------------------------------------------------------------------------------------------------|
| `test` | Runs all `test:*` scripts. |
| `test:unit` | Runs all [Jest] Unit Tests. Accepts Jest flags. |
| `test:lint` | Runs [ESLint] [rules](https://github.com/americanexpress/eslint-config-amex) against source. |
| `pretest:integration` | Using [`concurrently`], runs `build:prod-sample` and pulls Docker containers to prepare the [Integration Test Suite]. |
| `test:integration` | Executes the [Integration Test Suite] using [Jest]. |
| `test:danger` | Runs the [DangerJS tool](https://danger.systems/js/) that returns bundle size stats and a preflight checklist for PRs. |
| `test:git-history` | Validates the Git commit message format using the [`commitlint`](https://github.com/conventional-changelog/commitlint) tool. |
| `test:lockfile` | Validates that the `package-lock.json` file contains valid NPM registry URLs using the [`lockfile-lint`](https://github.com/lirantal/lockfile-lint) tool. |

**üìò More Information**
* Library: [`one-app-dev-cdn`]
* Library: [`one-app-dev-proxy`]
* [Development Tools API Docs](./Development-Tools.md)

[‚òùÔ∏è Return To Top](#Cli-Commands)

[`concurrently`]: https://www.npmjs.com/package/concurrently
[Babel]: https://babeljs.io/
[One App Bundler]: https://github.com/americanexpress/one-app-cli/tree/master/packages/one-app-bundler
[Webpack]: https://webpack.js.org/
[`prod-sample/sample-modules`]: ../../../prod-sample/sample-modules
[Integration Tests]: #test-commands
[`one-app-dev-proxy`]: https://github.com/americanexpress/one-app-dev-proxy
[development proxy]: https://github.com/americanexpress/one-app-dev-proxy
[`one-app-dev-cdn`]: https://github.com/americanexpress/one-app-dev-cdn
[development CDN]: https://github.com/americanexpress/one-app-dev-cdn
[Holocron Module]: ../README.md#modules
[Holocron Modules]: ../README.md#modules
[Jest]: https://jestjs.io/
[ESLint]: https://eslint.org/
[Integration Test Suite]: ../../../__tests__/integration/README.md
[Express Middleware]: http://expressjs.com/en/guide/writing-middleware.html#writing-middleware-for-use-in-express-apps