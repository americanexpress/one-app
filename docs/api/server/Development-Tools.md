[👈 Return to Overview](../README.md)

# Development Tools

Documentation Forthcoming

[☝️ Return To Top](#development-tools)