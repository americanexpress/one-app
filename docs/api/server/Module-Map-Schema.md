[üëà Return to Overview](../API.md)

# Module Map Schema

[**Holocron Modules**](../API.md#modules) may be developed and versioned in isolation as if they were their own frontend applications. The `one-app` Server relies on a `module-map.json` configuration file to orchestrate all these versioned web experiences together to form a single application. A Module Map contains an object of the following information:
* Holocron Module Name
* URLs to Holocron Module Bundles
* [Subresource Integrity](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity) Hash to verify contents of the Holocron Module Bundle

**Contents**
* [Shape](#shape)
* [Example](#example)
* [Field Annotations](#field-annotations)
  * [`key` Field](#key-field)
  * [`[moduleName]` Field](#moduleName-field)
  * [`browser` Field](#browser-field)
  * [`legacyBrowser` Field](#legacyBrowser-field)
  * [`node` Field](#node-field)
  * [`url` Field](#url-field)
  * [`integrity` Field](#integrity-field)

## Shape

```js
({
  key: String, // optional
  modules: {
    [moduleName]: { // required
      browser: {
        url: String, // required
        integrity: String, // SRI String required
      },
      legacyBrowser: {
        url: String, // required
        integrity: String, // SRI String required
      },
      node: {
        url: String, // required
        integrity: String, // SRI String required
      },
    },
    // ... more module entries allowed
  },
});
```

## Example
```json
{
  "key": "<any-string>",
  "modules": {
    "cultured-frankie": {
      "browser": {
        "url": "https://one-app-statics.surge.sh/modules/f16a872/cultured-frankie/0.0.0/cultured-frankie.browser.js",
        "integrity": "sha256-CGgTIe7x19CK+Z7G7YAjwPMstd/gpWNd4FRce7csbNA= sha384-8QStyVD+275LwFA2Zlgyh3fnt5chqPMpJjD9Yn8AF1dybM297+lFid40gZ+JCBqY"
      },
      "legacyBrowser": {
        "url": "https://one-app-statics.surge.sh/modules/f16a872/cultured-frankie/0.0.0/cultured-frankie.legacy.browser.js",
        "integrity": "sha256-7xmKAgHKfrK4WVdRDW5i2NHauPd79NWa/XJ5yKiKhpo= sha384-MsQDrDBAK3KfG/svpk7C3wHy+Ht7G6D99NbjZjBcfxdhPbZFeIQgsBtKkCVXzR1P"
      },
      "node": {
        "url": "https://one-app-statics.surge.sh/modules/f16a872/cultured-frankie/0.0.0/cultured-frankie.node.js",
        "integrity": "sha256-fQT5PYOExfnv/b6tMAoWdp8lyJK1XPHloKfjNQUfySk= sha384-LhjPddGOg6KcGbs2IhVWvZRnyNrEasr01aVhOPXmwlTzg3n2QQqxJHdx93aucRby"
      }
    }
  }
}
```

## Field Annotations

### `key` Field

This required key is used to bust the caching of Modules in the [Holocron Module Registry].

### `[moduleName]` Field

By convention, the `moduleName` is the key mapping where Holocron Module bundles are stored in the [Holocron Module Registry].

### `browser` Field

Includes an object of `url` and `integrity` pointing to the location of the modern `browser` bundle (e.g. last two major versions of currently used browsers) for a [Holocron Module].

### `legacyBrowser` Field

Includes an object of `url` and `integrity` pointing to the `legacyBrowser` bundle (e.g. more transpilation for older browsers like Internet Explorer) for a [Holocron Module].

### `node` Field

Includes an object of `url` and `integrity` pointing to the location of the `node` bundle (e.g. Serverside compatible) for a [Holocron Module].

### `url` Field

Contains a fully qualified `url` to a [Holocron Module] bundle location.

### `integrity` Field

Contains a hash [SRI](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity) `integrity` string to validate the contents of a [Holocron Module] bundle.

**üìò More Information**
* Library: [One App Bundler](https://github.com/americanexpress/one-app-cli/tree/master/packages/one-app-bundler)
* CLI Command: [npm start](./CLI-Commands.md#start-usage)
* [Environment Variables](./Environment-Variables.md)
* [Development Tools](./Development-Tools.md)

[‚òùÔ∏è Return To Top](#module-map-schema)

[Holocron Module Registry]: https://github.com/americanexpress/holocron/blob/master/packages/holocron/API.md#module-registry
[Holocron Module]: ../API.md#modules
