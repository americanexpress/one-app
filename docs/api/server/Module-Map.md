[👈 Return to Overview](../API.md)

# Module Map

Documentation Forthcoming

[☝️ Return To Top](#module-map)