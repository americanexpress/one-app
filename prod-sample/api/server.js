/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

const fs = require('fs');
const https = require('https');
const http = require('http');
const jsonServer = require('json-server');
const pause = require('connect-pause');

const server = jsonServer.create();
const router = jsonServer.router('db.json');
const middlewares = jsonServer.defaults();

const delay = process.argv[2] ? Number(process.argv[2]) : 0;
const options = {
  key: fs.readFileSync('./api-privkey.pem'),
  cert: fs.readFileSync('./api-cert.pem'),
};

if (delay !== 0) {
  server.use(pause(delay));
}

server.use(middlewares);
server.use((req, res, next) => {
  const secretMessage = req.headers['auth-token'] ? 'you are being watched' : 'unauthorised';
  res.append('secret-message', secretMessage);
  res.append('traceparent', req.headers.traceparent);
  next();
});
server.use(router);

https.createServer(options, server).listen(443, '0.0.0.0');
http.createServer({}, server).listen(80, '0.0.0.0');
