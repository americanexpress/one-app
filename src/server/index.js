/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import path from 'path';
import fs from 'fs';
import Intl from 'lean-intl';
import enData from 'lean-intl/locale-data/json/en.json';

import './init';
// Allow env config to run before importing holocron. (x2)
import ssrServer from './ssrServer';
import metricsServer from './metricsServer';
import listen from './listen';
import { addServer, shutdown } from './shutdown';
import pollModuleMap from './utils/pollModuleMap';
import loadModules from './utils/loadModules';

function ssrServerStart() {
  // need to load _some_ locale so that react-intl does not prevent modules from loading
  // eslint-disable-next-line no-underscore-dangle
  Intl.__addLocaleData(enData);
  return loadModules()
    .then(() => new Promise((res, rej) => {
      addServer(listen(ssrServer, (err, { port }) => {
        if (err) {
          rej(err);
        } else {
          console.log(`ðŸŒŽ One App server listening on port ${port}`);
          res();
        }
      }));
    }))
    .then(pollModuleMap);
}

function metricsServerStart() {
  return new Promise((res, rej) => {
    const port = process.env.HTTP_METRICS_PORT;
    addServer(metricsServer.listen(port, (err) => (err ? rej(err) : res(port))));
  })
    .then(
      (port) => console.log(`ðŸ“Š Metrics server listening on port ${port}`),
      (err) => {
        console.error('error encountered starting the metrics server', err);
        throw err;
      }
    );
}

function appServersStart() {
  return Promise.all([
    ssrServerStart(),
    metricsServerStart(),
  ]);
}

let serverChain;
if (process.env.NODE_ENV === 'development') {
  const pathToDevEndpoints = path.join(process.cwd(), '.dev', 'endpoints', 'index.js');

  const getRemotesFromDevEndpointsFile = () => {
    const moduleRemotes = {};
    if (fs.existsSync(pathToDevEndpoints)) {
      // eslint-disable-next-line global-require,import/no-dynamic-require
      Object.values(require(pathToDevEndpoints)()).forEach(({ destination, devProxyPath }) => {
        moduleRemotes[devProxyPath] = destination;
      });
      return moduleRemotes;
    }
    return {};
  };

  /* eslint
    global-require: 0,
    import/no-extraneous-dependencies: ["error", {"devDependencies": true }]
  */
  const { argv } = require('yargs');

  const watchLocalModules = require('./utils/watchLocalModules').default;
  const devHolocronCDN = require('./devHolocronCDN').default;
  const oneAppDevProxy = require('@americanexpress/one-app-dev-proxy');
  const oneAppDevCdnPort = process.env.HTTP_ONE_APP_DEV_CDN_PORT;
  const oneAppDevProxyPort = process.env.HTTP_ONE_APP_DEV_PROXY_SERVER_PORT;

  serverChain = Promise.resolve()
    .then(() => new Promise((res, rej) => addServer(
      devHolocronCDN.listen(oneAppDevCdnPort, (err) => {
        if (err) {
          rej(err);
        } else {
          console.log(`ðŸ‘• one-app-dev-cdn server listening on port ${oneAppDevCdnPort}`);
          res();
        }
      })
    )))
    .then(() => new Promise((res, rej) => addServer(
      oneAppDevProxy({
        useMiddleware: argv.m,
        pathToMiddleware: path.join(__dirname, '../../.dev/middleware'),
        remotes: getRemotesFromDevEndpointsFile(),
      }).listen(oneAppDevProxyPort, (err) => {
        if (err) {
          rej(err);
        } else {
          console.log(`ðŸ‘– one-app-dev-proxy server listening on port ${oneAppDevProxyPort}`);
          res();
        }
      })
    )))
    .then(appServersStart)
    .then(watchLocalModules);
} else {
  serverChain = appServersStart();
}

export default serverChain.catch((err) => {
  console.error(err);
  shutdown();
});
