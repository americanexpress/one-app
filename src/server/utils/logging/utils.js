/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { STATUS_CODES } from 'http';

import chalk from 'chalk';

const coloredLevels = {
  error: chalk.red('error'),
  warn: chalk.yellow('warn'),
  log: chalk.blue('log'),
  info: chalk.gray('info'),
};

function printStatusCode(obj) {
  const { request: { statusCode } } = obj;

  if (statusCode === null) {
    return chalk.black.bgRed(statusCode);
  }

  if (statusCode < 300) {
    // 100s, 200s
    return chalk.green(statusCode);
  }
  if (statusCode < 400) {
    // 300s
    return chalk.blue(statusCode);
  }
  if (statusCode < 500) {
    // 400s
    return chalk.yellow(statusCode);
  }
  // 500s
  return chalk.red(statusCode);
}

function printStatusMessage(obj) {
  const { request: { statusText, statusCode } } = obj;

  return statusText || STATUS_CODES[statusCode] || chalk.black.bgRed('Timed Out?');
}

function printDurationTime(obj) {
  const { request: { timings: { duration } } } = obj;
  if (duration < 100) {
    return chalk.green(duration);
  }
  if (duration < 1e3) {
    return chalk.yellow(duration);
  }
  if (duration < 6e3) {
    return chalk.red(duration);
  }
  return chalk.black.bgRed(duration);
}

export {
  coloredLevels,
  printStatusCode,
  printStatusMessage,
  printDurationTime,
};
