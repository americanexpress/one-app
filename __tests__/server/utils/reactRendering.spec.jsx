/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { renderToString, renderToStaticMarkup } from 'react-dom/server';
import { Helmet } from 'react-helmet';

import { renderForString, renderForStaticMarkup } from '../../../src/server/utils/reactRendering';

describe('reactRendering', () => {
  describe('renderForString', () => {
    it('is a function', () => expect(renderForString).toBeInstanceOf(Function));

    it('returns an object with "renderedString" and "helmetInfo" keys', () => {
      const HelloComponent = () => <p>Hello</p>;
      expect(renderForString(<HelloComponent />)).toEqual({
        renderedString: expect.any(String),
        helmetInfo: expect.any(Object),
      });
    });

    it('calls react-dom renderToString', () => {
      const HelloComponent = () => <p>Hello</p>;
      const renderResults = renderForString(<HelloComponent />);
      expect(renderResults).toHaveProperty('renderedString');
      expect(renderResults.renderedString).toMatchSnapshot('<p>Hello</p>');
    });

    it('does not use existing Helmet data for the render', () => {
      const PreviousRender = () => (
        <div>
          <Helmet
            {...{
              htmlAttributes: {
                lang: 'en-NZ',
                previous: 'render',
              },
              title: 'Page Title of Previous Render',
            }}
          />
          <h1>previous title</h1>
          <p>previous content</p>
        </div>
      );
      renderToString(<PreviousRender />);

      const IntendedRender = () => (
        <div>
          <Helmet
            {...{
              htmlAttributes: {
                lang: 'en-GB',
                intended: 'render',
              },
              title: 'Page Title of Intended Render',
            }}
          />
          <h1>intended title</h1>
          <p>intended content</p>
        </div>
      );
      const { helmetInfo } = renderForString(<IntendedRender />);

      expect(helmetInfo.title.toString()).toMatchSnapshot();
      expect(helmetInfo.htmlAttributes.toString()).toMatchSnapshot();
    });

    it('provides Helmet data of just the render', () => {
      const IntendedRender = () => (
        <div>
          <Helmet
            {...{
              htmlAttributes: {
                lang: 'en-GB',
                intended: 'render',
              },
              title: 'Page Title of Intended Render',
            }}
          />
          <h1>intended title</h1>
          <p>intended content</p>
        </div>
      );
      const { helmetInfo } = renderForString(<IntendedRender />);

      const NextRender = () => (
        <div>
          <Helmet
            {...{
              htmlAttributes: {
                lang: 'en-NZ',
                next: 'render',
              },
              title: 'Page Title of Next Render',
            }}
          />
          <h1>next title</h1>
          <p>next content</p>
        </div>
      );
      renderToString(<NextRender />);

      expect(helmetInfo.title.toString()).toMatchSnapshot();
      expect(helmetInfo.htmlAttributes.toString()).toMatchSnapshot();
    });
  });
  describe('renderForStaticMarkup', () => {
    it('is a function', () => expect(renderToStaticMarkup).toBeInstanceOf(Function));

    it('returns an object with "renderedString" and "helmetInfo" keys', () => {
      const HelloComponent = () => <p>Hello</p>;
      expect(renderForStaticMarkup(<HelloComponent />)).toEqual({
        renderedString: expect.any(String),
        helmetInfo: expect.any(Object),
      });
    });

    it('calls react-dom renderToStaticMarkup', () => {
      const HelloComponent = () => <p>Hello</p>;
      const renderResults = renderForStaticMarkup(<HelloComponent />);
      expect(renderResults).toHaveProperty('renderedString');
      expect(renderResults.renderedString).toMatchSnapshot('<p>Hello</p>');
    });

    it('does not use existing Helmet data for the render', () => {
      const PreviousRender = () => (
        <div>
          <Helmet
            {...{
              htmlAttributes: {
                lang: 'en-NZ',
                previous: 'render',
              },
              title: 'Page Title of Previous Render',
            }}
          />
          <h1>previous title</h1>
          <p>previous content</p>
        </div>
      );
      renderToStaticMarkup(<PreviousRender />);

      const IntendedRender = () => (
        <div>
          <Helmet
            {...{
              htmlAttributes: {
                lang: 'en-GB',
                intended: 'render',
              },
              title: 'Page Title of Intended Render',
            }}
          />
          <h1>intended title</h1>
          <p>intended content</p>
        </div>
      );
      const { helmetInfo } = renderForStaticMarkup(<IntendedRender />);

      expect(helmetInfo.title.toString()).toMatchSnapshot();
      expect(helmetInfo.htmlAttributes.toString()).toMatchSnapshot();
    });

    it('provides Helmet data of just the render', () => {
      const IntendedRender = () => (
        <div>
          <Helmet
            {...{
              htmlAttributes: {
                lang: 'en-GB',
                intended: 'render',
              },
              title: 'Page Title of Intended Render',
            }}
          />
          <h1>intended title</h1>
          <p>intended content</p>
        </div>
      );
      const { helmetInfo } = renderForStaticMarkup(<IntendedRender />);

      const NextRender = () => (
        <div>
          <Helmet
            {...{
              htmlAttributes: {
                lang: 'en-NZ',
                next: 'render',
              },
              title: 'Page Title of Next Render',
            }}
          />
          <h1>next title</h1>
          <p>next content</p>
        </div>
      );
      renderToStaticMarkup(<NextRender />);

      expect(helmetInfo.title.toString()).toMatchSnapshot();
      expect(helmetInfo.htmlAttributes.toString()).toMatchSnapshot();
    });
  });
});
