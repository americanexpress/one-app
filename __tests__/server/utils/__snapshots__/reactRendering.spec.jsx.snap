// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reactRendering renderForStaticMarkup calls react-dom renderToStaticMarkup: <p>Hello</p> 1`] = `"<p>Hello</p>"`;

exports[`reactRendering renderForStaticMarkup does not use existing Helmet data for the render 1`] = `"<title data-react-helmet="true">Page Title of Intended Render</title>"`;

exports[`reactRendering renderForStaticMarkup does not use existing Helmet data for the render 2`] = `"lang="en-GB" intended="render""`;

exports[`reactRendering renderForStaticMarkup provides Helmet data of just the render 1`] = `"<title data-react-helmet="true">Page Title of Intended Render</title>"`;

exports[`reactRendering renderForStaticMarkup provides Helmet data of just the render 2`] = `"lang="en-GB" intended="render""`;

exports[`reactRendering renderForString calls react-dom renderToString: <p>Hello</p> 1`] = `"<p data-reactroot="">Hello</p>"`;

exports[`reactRendering renderForString does not use existing Helmet data for the render 1`] = `"<title data-react-helmet="true">Page Title of Intended Render</title>"`;

exports[`reactRendering renderForString does not use existing Helmet data for the render 2`] = `"lang="en-GB" intended="render""`;

exports[`reactRendering renderForString provides Helmet data of just the render 1`] = `"<title data-react-helmet="true">Page Title of Intended Render</title>"`;

exports[`reactRendering renderForString provides Helmet data of just the render 2`] = `"lang="en-GB" intended="render""`;
